domain/agentbinary/state/state.go

```diff
@@ -6,28 +6,140 @@ package state
 	"github.com/juju/juju/domain"
 	"github.com/juju/juju/domain/agentbinary"
+	agentbinaryerrors "github.com/juju/juju/domain/agentbinary/errors"
+	jujudb "github.com/juju/juju/internal/database"
+	"github.com/juju/juju/internal/errors"
 )
 
 // State represents a type for interacting with the underlying state.
 type State struct {
 	*domain.StateBase
 }
 
-// NewState returns a new State for interacting with the underlying state.
+// NewState returns a new State for interacting with agent binaries stored in
+// the database.
 func NewState(factory database.TxnRunnerFactory) *State {
 	return &State{
 		StateBase: domain.NewStateBase(factory),
 	}
 }
 
 // Add adds a new agent binary's metadata to the database.
-// It always overwrites the metadata for the given version and arch if it already exists.
-// It returns [coreerrors.NotSupported] if the architecture is not found in the database.
-// It returns [coreerrors.NotFound] if object store UUID is not found in the database.
+// [agentbinaryerrors.AlreadyExists] when the provided agent binary already
+// exists.
+// [agentbinaryerrors.ObjectNotFound] when no object exists that matches
+// this agent binary.
+// [coreerrors.NotSupported] if the architecture is not supported by the
+// state layer.
 func (s *State) Add(ctx context.Context, metadata agentbinary.Metadata) error {
-	// TODO: implement this method.
+	db, err := s.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	archVal := architectureRecord{Name: metadata.Arch}
+	agentBinary := agentBinaryRecord{
+		Version:         metadata.Version,
+		ObjectStoreUUID: metadata.ObjectStoreUUID.String(),
+	}
+
+	archStmt, err := s.Prepare(`
+SELECT &architectureRecord.*
+FROM architecture
+WHERE name = $architectureRecord.name
+`, archVal)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	insertStmt, err := s.Prepare(`
+INSERT INTO agent_binary_store (*) VALUES ($agentBinaryRecord.*)
+`, agentBinary)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Check if the object exists in the database for RI.
+		exists, err := s.checkObjectExists(ctx, tx, metadata.ObjectStoreUUID)
+		if err != nil {
+			return errors.Capture(err)
+		}
+		if !exists {
+			return errors.Errorf(
+				"object with id %q does not exist in store",
+				metadata.ObjectStoreUUID,
+			).Add(agentbinaryerrors.ObjectNotFound)
+		}
+
+		// Check if the architecture exists and get its ID
+		err = tx.Query(ctx, archStmt, archVal).Get(&archVal)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf(
+				"architecture %q is not supported",
+				metadata.Arch,
+			).Add(coreerrors.NotSupported)
+		} else if err != nil {
+			return errors.Capture(err)
+		}
+
+		agentBinary.ArchitectureID = archVal.ID
+
+		err = tx.Query(ctx, insertStmt, agentBinary).Run()
+		if jujudb.IsErrConstraintPrimaryKey(err) {
+			return errors.New(
+				"agent binary already exists",
```

Comment:
Better to include the version and arch in the error message.

Reply:
Also, we can add a comment to explain why we got this:
```
// There must be a binary with a different SHA but the same version and arch.
```

Reply:
Done & Done